spring-cloud-dataflow-core/src/main/java/org/springframework/cloud/dataflow/core/StreamApplicationDefinitionBuilder.java

```diff
@@ -85,19 +88,48 @@ else if (argument.getName().equalsIgnoreCase("outputType")) {
 					}
 				}
 			}
-			if (m > 0) {
-				builder.setProperty(BindingPropertyKeys.INPUT_DESTINATION,
-						String.format("%s.%s", streamName, appNodes.get(m - 1).getLabelName()));
-				builder.setProperty(BindingPropertyKeys.INPUT_GROUP, streamName);
+			if (appNode.isLongLivedNonStreamApp()) {
+				builder.setApplicationType(ApplicationType.app);
 			}
-			if (m < appNodes.size() - 1) {
-				builder.setProperty(BindingPropertyKeys.OUTPUT_DESTINATION,
-						String.format("%s.%s", streamName, appNode.getLabelName()));
-				builder.setProperty(BindingPropertyKeys.OUTPUT_REQUIRED_GROUPS, streamName);
+			else {
+				if (m == 0) {
+					if (sourceDestination == null) {
+						builder.setApplicationType(ApplicationType.source);
+					}
+					else {
+						if (appNodes.size() == 1 && sinkDestination == null) {
+							builder.setApplicationType(ApplicationType.sink);
+						}
+						else {
+							builder.setApplicationType(ApplicationType.processor);
+						}
+					}
+				}
+				else {
+					if (m < appNodes.size() - 1 || sinkDestination != null) {
+						builder.setApplicationType(ApplicationType.processor);
+					}
+					else {
+						builder.setApplicationType(ApplicationType.sink);
+					}
+
+				}
+			}
+			// Add binding property keys only if the app is a source, processor or sink.
```

Comment:
We still set some of the binding properties in other places (like contentType, consumer group) properties. I miss understanding what binding properties should be set and what not.